src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -319,50 +368,116 @@ public WPI_Pigeon2 getGyro() {
-            // Optional, defaults to true
-            this // Requires this drive subsystem
-            ),
-        new InstantCommand(
-            () -> {
-              targetHeadingDegrees = getPose().getRotation().getDegrees();
-            }));
+            new PrintCommand("Finished a trajectory"))
+        .withName("Follow trajectory");
+  }
+
+  public void setLimelightTargetFromTransform(Transform2d transform, double latencySec) {
+    // Transform is to get the limelight to the correct location, not to get the robot
+    // Here we correct for that
+    Transform2d flipTransform =
+        new Transform2d(
+            new Translation2d(-transform.getX(), transform.getY()), transform.getRotation());
+    System.out.println("Flip Transform: " + flipTransform.getX() + " " + flipTransform.getY());
+
+    Pose2d curPose = getPose();
+    latencySec = latencySec + 0.05;
+    Transform2d pastTransform =
+        new Transform2d(
+                new Translation2d(
+                    lastSetChassisSpeeds.vxMetersPerSecond * latencySec,
+                    lastSetChassisSpeeds.vyMetersPerSecond * latencySec),
+                Rotation2d.fromRadians(lastSetChassisSpeeds.omegaRadiansPerSecond * latencySec))
+            .inverse();
+    Pose2d pastPose = curPose.plus(pastTransform);
+
+    final boolean useLatencyAdjustment = true;
+
+    targetPose =
+        useLatencyAdjustment
+            ? Optional.of(pastPose.plus(flipTransform))
+            : Optional.of(curPose.plus(flipTransform));
+  }
+
+  public PathPlannerTrajectory poseToPath(boolean red) {
+    Pose2d curPose = getPose();
+    double coastLatencySec = 0.03;
+    Transform2d coastTransform =
+        new Transform2d(
+            new Translation2d(
+                lastSetChassisSpeeds.vxMetersPerSecond * coastLatencySec,
+                lastSetChassisSpeeds.vyMetersPerSecond * coastLatencySec),
+            Rotation2d.fromRadians(lastSetChassisSpeeds.omegaRadiansPerSecond * coastLatencySec));
+    Pose2d futurePose = curPose.plus(coastTransform);
+
+    System.out.println("Acquired target? " + targetPose.isPresent());
+    if (targetPose.isPresent()) {
+      Transform2d trajectoryTransform = targetPose.get().minus(futurePose);
+      System.out.println(
+          "Trajectory Transform: " + trajectoryTransform.getX() + " " + trajectoryTransform.getY());
+    }
+    Pose2d finalPose =
+        targetPose.isPresent()
+            ? targetPose.get()
+            : futurePose.plus(
+                new Transform2d(
+                    new Translation2d(-1.5, red ? 1.15 : -1.15), Rotation2d.fromDegrees(0)));
+    Transform2d finalTransform =
+        new Transform2d(finalPose.getTranslation(), finalPose.getRotation());
+    // Rotation2d startHeading =
+    // moveTransform.getTranslation().getAngle().plus(curPose.getRotation());
+    // Rotation2d finalHeading = startHeading.plus(Rotation2d.fromDegrees(180));
```

Comment:
Commented code should probably be removed

Reply:
Removed